### PR TITLE
[Dynamo] Make torch.autograd.Function.forward support graph break and no re-compilation

### LINKED ISSUE #91101: [Dynamo] Graph Re-compilation Invoked by `np.sqrt`
### 🐛 Describe the bug

We find that `np.sqrt` can lead to graph re-compilation, but `math.sqrt` will not. I have brief search in the Github issues but didn't find realted issue, so I think this might be one bug.

Here's one minimal example to reproduce:

```python
#!/usr/bin/env python

import sys
import torch
from torch._dynamo import optimize, config
config.cache_size_limit = 4
torch.manual_seed(0)

class Op(torch.autograd.Function):
    @staticmethod
    def forward(ctx, i):
        result = i * sqrt(i.numel())
        ctx.save_for_backward(result)
        return result

    @staticmethod
    def backward(ctx, grad_output):
        result, = ctx.saved_tensors
        return grad_output * sqrt(result.numel())

@optimize("inductor")
def toy_example(x):
    return Op.apply(x).norm()

def test():
    for i in range(5):
        x = torch.randn(10)
        print(toy_example(x))

if __name__ == "__main__":
    msg = "Usage: python bug-2.py numpy|math"
    assert len(sys.argv) == 2, msg
    if sys.argv[1] == "numpy":
        from numpy import sqrt
    elif sys.argv[1] == "math":
        from math import sqrt
    else:
        raise  Exception(msg)
    test()
```

Here's the message I got with PyTorch `1.14.0a0+410ce96`:

```bash
$ python bug-2.py numpy
tensor(11.0491)
tensor(8.3844)
tensor(9.0465)
tensor(9.7683)
[2022-12-19 14:51:04,617] torch._dynamo.convert_frame: [WARNING] torch._dynamo hit config.cache_size_limit (4)
   function: '<graph break in forward>' (bug-2.py:12)
   reasons:  ['___guarded_code.valid']
to diagnose recompilation issues, see https://github.com/pytorch/torchdynamo/blob/main/TROUBLESHOOTING.md.
tensor(8.0597)

$ python bug-2.py math
tensor(11.0491)
tensor(8.3844)
tensor(9.0465)
tensor(9.7683)
tensor(8.0597)
```

### Versions

```
Collecting environment information...
PyTorch version: 1.14.0a0+410ce96
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.24.1
Libc version: glibc-2.31

Python version: 3.8.10 (default, Jun 22 2022, 20:18:18)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-99-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-80GB
Nvidia driver version: 515.65.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.2
[pip3] open-clip-torch==2.7.0
[pip3] pytorch-quantization==2.1.2
[pip3] torch==1.14.0a0+410ce96
[pip3] torch-tensorrt==1.3.0a0
[pip3] torchtext==0.13.0a0+fae8e8c
[pip3] torchvision==0.15.0a0
[conda] Could not collect
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @mlazos @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @jjsjann123 @kevinstephano

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #91101

cc @mlazos @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -872,7 +872,7 @@ def test_reformer_train(self):
         with torch.enable_grad():
             cnt = self._reformer(nopython=False)
         # cant inline torch.autograd.Function means graph break
-        self.assertEqual(cnt.frame_count, 4)
+        self.assertEqual(cnt.frame_count, 3)
```
Comment (yanboliang) [inline test/dynamo/test_repros.py:875]: No graph break on ```ctx.save_for_backward```, so save one frame count. Actually this unit test triggers re-complication before this PR, but we didn't catch it.

**Review by jansel — APPROVED** [review]

File: test/dynamo/test_repros.py (hunk 1 above)
Comment (mlazos) [reply test/dynamo/test_repros.py:875]: Yeah I've noticed frame count is really not the best way of detecting these. I started using explain() to count the graphs and graph breaks

**Comment by ngimel** [comment]
@yanboliang can we land this?

**Review by mlazos — APPROVED** [review]

**Comment by ngimel** [comment]
But apparently even with this fix the recompilation problem with the custom autograd function persists, see https://github.com/pytorch/pytorch/issues/91101#issuecomment-1378425130, can we propose some fixes? @mlazos is it possible to call `allow_in_graph` to avoid graph breaks in those cases?

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -464,6 +465,9 @@ def index_source(key):
             return AutogradFunctionVariable(
                 value, guards=make_guards(GuardBuilder.FUNCTION_MATCH)
             )
+        elif isinstance(value, torch.autograd.function.FunctionCtx):
+            # The autograd.function context for forward extraction
+            return BlackHoleVariable()
```
Comment (ngimel) [inline torch/_dynamo/variables/builder.py:470]: This won't work if you need to extract `saved_tensors` in backward?

Comment (yanboliang) [reply torch/_dynamo/variables/builder.py]: Yes, so I wrap it as a ```AutogradFunctionContextVariable``` to solve this problem, the behavior is falling back if users trying to access saved tensors. ```allow_in_graph``` is the workaround way, it skips tracing for the forward/backward functions completely, so it may loss dynamo optimization opportunity.
